File: neet_code_all_cpp/two_pointers/reverse_string.cpp
#include <iostream>
#include <vector>

class Solution {
    // AC
    // 模範解答も全く同じ
    public:
        void reverseString(std::vector<char>& s) {
            auto [l, r] = std::pair(0, s.size() - 1);
            while(l < r) {
                std::swap(s[l], s[r]);
                ++l;
                --r;
            }
        }
};

void show_result(std::vector<char>& result) {
    for(auto&c :result) {
        std::cout << c << " ";
    }
    std::cout << std::endl;
}

int main(void) {
    std::vector<char> case_1 = {'h', 'e', 'l', 'l', 'o'};
    std::vector<char> case_2 = {'H', 'a', 'n', 'n', 'a', 'h'};

    Solution s_1;
    auto res_1 = case_1;
    s_1.reverseString(res_1);
    auto res_2 = case_2;
    s_1.reverseString(res_2);

    show_result(res_1);
    show_result(res_2);
}

File: neet_code_all/src/two_pointers/reverse_string.rs
struct Solution {}
impl Solution {
    // AC
    // 模範解答は全く同じ
    pub fn reverse_string(s: &mut Vec<char>) {
        let (mut l, mut r) = (0, s.len() - 1);
        while l < r {
            s.swap(l, r);
            l += 1;
            r -= 1;
        }
    }
}
fn main() {
    let case_1 = vec!['h', 'e', 'l', 'l', 'o'];
    // => ['o','l','l','e','h']
    let case_2 = vec!['H', 'a', 'n', 'n', 'a', 'h'];
    // => ['h', 'a', 'n', 'n', 'a', 'H']

    let mut res_1 = case_1.clone();
    Solution::reverse_string(&mut res_1);
    println!("case_1: {:?}", res_1);

    let mut res_2 = case_2.clone();
    Solution::reverse_string(&mut res_2);
    println!("case_2: {:?}", res_2);
}
